tests/naive.py

```diff
@@ -1388,55 +1412,64 @@ def aampdist_snippets(
     )
 
 
-def prescrump(T_A, m, T_B, s, exclusion_zone=None):
+def prescrump(T_A, m, T_B, s, exclusion_zone=None, k=1):
     dist_matrix = distance_matrix(T_A, T_B, m)
 
     n_A = T_A.shape[0]
     l = n_A - m + 1
 
-    P = np.empty(l)
-    I = np.empty(l, dtype=np.int64)
-    P[:] = np.inf
-    I[:] = -1
+    P = np.full((l, k), np.inf, dtype=np.float64)
+    I = np.full((l, k), -1, dtype=np.int64)
 
     for i in np.random.permutation(range(0, l, s)):
         distance_profile = dist_matrix[i]
         if exclusion_zone is not None:
             apply_exclusion_zone(distance_profile, i, exclusion_zone, np.inf)
 
-            # only for self-join
-            mask = distance_profile < P
-            P[mask] = distance_profile[mask]
-            I[mask] = i
-
-        I[i] = np.argmin(distance_profile)
-        P[i] = distance_profile[I[i]]
-        if P[i] == np.inf:
-            I[i] = -1
-        else:
-            j = I[i]
-            for k in range(1, min(s, l - max(i, j))):
-                d = dist_matrix[i + k, j + k]
-                if d < P[i + k]:
-                    P[i + k] = d
-                    I[i + k] = j + k
-                if d < P[j + k]:
-                    P[j + k] = d
-                    I[j + k] = i + k
-
-            for k in range(1, min(s, i + 1, j + 1)):
-                d = dist_matrix[i - k, j - k]
-                if d < P[i - k]:
-                    P[i - k] = d
-                    I[i - k] = j - k
-                if d < P[j - k]:
-                    P[j - k] = d
-                    I[j - k] = i - k
+        I[i, 1:] = I[i, :-1]
+        I[i, 0] = np.argmin(distance_profile)
+        P[i, 1:] = P[i, :-1]
+        P[i, 0] = distance_profile[I[i, 0]]
+
+        if P[i, 0] == np.inf:
+            I[i, 0] = -1
+            continue
+
+        j = I[i, 0]  # index of 1st NN
+        for g in range(1, min(s, l - max(i, j))):
+            d = dist_matrix[i + g, j + g]
+            if d < P[i + g, -1]:
+                pos = np.searchsorted(P[i + g], d, side="right")
+                P[i + g] = np.insert(P[i + g], pos, d)[:-1]
+                I[i + g] = np.insert(I[i + g], pos, j + g)[:-1]
+            if d < P[j + g]:
```

Comment:
Let's add a test to catch this if possible if you think it is worth it. I'm also not sure why the tests pass but it likely means that we don't have the right inputs that would affect/trigger this OR the naive version also persists this

Reply:
It sounds like we should slow down a little and take a closer look at `scrump`/`prescrump`

Reply:
> Let's add a test to catch this if possible if you think it is worth it

The reason that I got confused is that I think `numpy` gives an error for this line `if d < P[j + g]:` but we got no error. That means both `for-loops` (that contain these if-blocks) are probably skipped in our tests!

Reply:
I think I found the cause of this behavior. In the `tests/test_scrump.py`, we can see the following lines:

```
m = 3
zone = int(np.ceil(m / 4))
for s in range(1, zone + 1):
    seed = np.random.randint(100000)

    np.random.seed(seed)
    ref_P, ref_I = naive.prescrump(T_B, m, T_B, s=s, exclusion_zone=zone)
```

with `m=3`, the value of `zone` is `1` and thus the `s` value in the for-loop is just `s=1`  (i.e. just one single iteration with `s=1`). 

Now, if we take a look at the `tests/naive.py: prescrump`, we can see:
```
for g in range(1, min(s, l - max(i, j))):
       ###

for g in range(1, min(s, i + 1, j + 1)):
       ###
```

Since `s = 1`, both of these for-loops are skipped! I think we might be able to solve this by using `m=5`. 

What do you think?

Reply:
That makes sense! Yes, please create a new, separate unit test with `m=5`. Although, if our unit tests never enters that those code lines, our coverage testing should've found it too. Hmmm

Reply:
So I checked out the test functions again and I think the test function "test_prescrump_self_join_larger_window" is the only one that supports larger `m` and thus `s>1`. Hence, it goes through the for loop. This test function is just for regular (1NN, k=1) case (I didn't add it for KNN k>1 case) So, we do not get any error even though we miss that `-1` in our if condition.

Reply:
So, would it make sense to add a `test_prescrump_self_join_larger_window_m_5_k_5` ? where `m=5` and `k=5`?

Reply:
Yeah...I think that should work! I will add this new test function.